File: python/examples/filter/filter_demo.py
"""
Copyright (c) 2024 DevAM. All Rights Reserved.

SPDX-License-Identifier: GPL-2.0-only
"""

import sys
import os
import platform
from typing import *

system: str = platform.system().lower()
architecture: str = platform.machine().lower()

lemonshark_directory: str = os.path.dirname(os.path.realpath(__file__)) + f"/../../lemonshark"
lemonshark_directory = os.path.realpath(lemonshark_directory)
sys.path.append(lemonshark_directory)

from lemonshark import LemonShark, Session, Filter

def main():
    wireshark_directory: str = None
    if system == "windows":
        if architecture == "amd64":
            wireshark_directory = os.path.dirname(os.path.realpath(__file__)) + f"/../../../build/wireshark/windows/amd64/run/RelWithDebInfo"
            wireshark_directory = os.path.realpath(wireshark_directory)

    LemonShark.init([wireshark_directory])

    trace_file_path: str = os.environ["LS_EXAMPLE_FILE"]

    session: Session = Session.create_from_file(trace_file_path, "")

    # Test a valid filter
    filter:str = "frame.len < 150"
    (is_valid, error_message) = Filter.is_valid(filter)
    if(is_valid):
        print(f"'{filter}' is valid.")
    else:
        print(f"'{filter}' is invalid: {error_message}")
        
    # Test an invalid filter
    filter = "frame.len ! 150"
    (is_valid, error_message) = Filter.is_valid(filter)
    if(is_valid):
        print(f"'{filter}' is valid.")
    else:
        print(f"'{filter}' is invalid: {error_message}")

    session.close()

File: liblemonshark/examples/filter/filter_demo.c
/*
Copyright (c) 2024 DevAM. All Rights Reserved.

SPDX-License-Identifier: GPL-2.0-only
*/

#include "lemonshark.h"

static void handle_error(const char *error_message)
{
    g_print("%s", error_message);
}

int main(void)
{
    void *error_handler = ls_error_handler_add(handle_error);

	// Set the environment variable LS_EXAMPLE_FILE to the path of a valid trace file before running this example.
    const char *trace_file_path = g_getenv("LS_EXAMPLE_FILE");

    char *error_message = NULL;
    gint32 init_result = ls_session_create_from_file(trace_file_path, "", NULL, &error_message);

    if (init_result == LS_ERROR)
    {
        g_print("%s", error_message);
        g_free(error_message);
        return -1;
    }

    // Test a valid filter
    const char *filter = "frame.len < 150";
    error_message = NULL;
    gint32 is_valid_filter = ls_filter_is_valid(filter, &error_message);
    if (is_valid_filter)
    {
        g_print("'%s' is valid\n", filter);
    }
    else
    {
        g_print("'%s' is invalid: %s\n", filter, error_message);
    }
    if (error_message != NULL)
    {
        g_free(error_message);
    }

    // Test an invalid filter
    filter = "frame.len ! 150";
    error_message = NULL;
    is_valid_filter = ls_filter_is_valid(filter, &error_message);

    if (is_valid_filter)
    {
        g_print("'%s' is valid\n", filter);
    }
    else
    {
        g_print("'%s' is invalid: %s\n", filter, error_message);
    }

    if (error_message != NULL)
    {
        g_free(error_message);
    }

    ls_session_close();

    ls_error_handler_remove(error_handler);

    return 0;
}